core/shared/src/main/scala/fs2/internal/FreeC.scala

```diff
@@ -12,138 +12,204 @@ import scala.util.control.NonFatal
+        Bind(fx.translate(f), (e: Result[Any]) => k(e).translate(f))
+      case Eval(fx)  => sys.error("impossible")
+      case Result(r) => r.asFreeC[G]
     }
   }
 }
 
 private[fs2] object FreeC {
-  final case class Pure[F[_], R](r: R) extends FreeC[F, R] {
-    override def translate[G[_]](f: F ~> G): FreeC[G, R] =
-      this.asInstanceOf[FreeC[G, R]]
-    override def toString: String = s"FreeC.Pure($r)"
+
+  def pure[F[_], A](a: A): FreeC[F, A] = Result.Pure(a)
+
+  def eval[F[_], A](f: F[A]): FreeC[F, A] = Eval(f)
+
+  def raiseError[F[_], A](rsn: Throwable): FreeC[F, A] = Result.Fail(rsn)
+
+  def interrupted[F[_], A](scopeId: Token, failure: Option[Throwable]): FreeC[F, A] =
+    Result.Interrupted(scopeId, failure)
+
+  sealed trait Result[+R] { self =>
+
+    def asFreeC[F[_]]: FreeC[F, R] = self.asInstanceOf[FreeC[F, R]]
+
+    def asExitCase: ExitCase[Throwable] = self match {
+      case Result.Pure(_)           => ExitCase.Completed
+      case Result.Fail(err)         => ExitCase.Error(err)
+      case Result.Interrupted(_, _) => ExitCase.Canceled
+    }
+
+    def covary[R2 >: R]: Result[R2] = self.asInstanceOf[Result[R2]]
+
+    def recoverWith[R2 >: R](f: Throwable => Result[R2]): Result[R2] = self match {
+      case Result.Fail(err) =>
+        try { f(err) } catch { case NonFatal(err2) => Result.Fail(CompositeFailure(err, err2)) }
+      case _ => covary[R2]
+    }
+
+  }
+
+  object Result {
+
+    val unit: Result[Unit] = pure(())
+
+    def pure[A](a: A): Result[A] = Result.Pure(a)
+
+    def raiseError[A](rsn: Throwable): Result[A] = Result.Fail(rsn)
+
+    def interrupted[A](scopeId: Token, failure: Option[Throwable]): Result[A] =
+      Result.Interrupted(scopeId, failure)
+
+    def fromEither[F[_], R](either: Either[Throwable, R]): Result[R] =
+      either.fold(Result.Fail(_), Result.Pure(_))
+
+    def unapply[F[_], R](freeC: FreeC[F, R]): Option[Result[R]] = freeC match {
+      case r @ Result.Pure(_)           => Some(r: Result[R])
+      case r @ Result.Fail(_)           => Some(r: Result[R])
+      case r @ Result.Interrupted(_, _) => Some(r: Result[R])
+      case _                            => None
+    }
+
+    final case class Pure[F[_], R](r: R) extends FreeC[F, R] with Result[R] {
+      override def translate[G[_]](f: F ~> G): FreeC[G, R] =
+        this.asInstanceOf[FreeC[G, R]]
+      override def toString: String = s"FreeC.Pure($r)"
+    }
+
+    final case class Fail[F[_], R](error: Throwable) extends FreeC[F, R] with Result[R] {
+      override def translate[G[_]](f: F ~> G): FreeC[G, R] =
+        this.asInstanceOf[FreeC[G, R]]
+      override def toString: String = s"FreeC.Fail($error)"
+    }
+
+    // finalizerErrors contains accumulated errors during finalization of interruption.
+    // they will be presented to freeC after interupted scope will be closed
+    final case class Interrupted[F[_], R](scope: Token, finalizerErrors: Option[Throwable])
```

Comment:
The terminology here scares me a little bit -- `scope` and `finalizer` are stream concerns, not `FreeC` concerns. Could you add some docs either here or at the class level that describes this interruption model a little bit?

Reply:
ok yeah. It is a bit confusing, unfortunatelly we had not find a better way. I'll think of something.